lib/protobuf/generators/field_generator.rb

```diff
@@ -101,15 +110,30 @@ def packed?
 
       # Determine the field type
       def type_name
-        @type_name ||= begin
-                         case descriptor.type.name
-                         when :TYPE_MESSAGE, :TYPE_ENUM, :TYPE_GROUP then
-                           modulize(descriptor.type_name)
-                         else
-                           type_name = descriptor.type.name.to_s.downcase.sub(/type_/, '')
-                           ":#{type_name}"
-                         end
-                       end
+        @type_name ||= determine_type_name(descriptor)
+      end
+
+      # If this field is a map field, this returns a message descriptor that
+      # represents the entries in the map. Returns nil if this field is not
+      # a map field.
+      def map_entry
+        @map_entry ||= determine_map_entry
+      end
+
+      def map?
+        !map_entry.nil?
+      end
+
+      def map_key_type_name
```

Comment:
Can we use a slightly more descriptive variable name than `e` for this since it's multi line please? :P Alternatively, I would just call it `map_entry` since you memoize that value